### PR TITLE
Move torch::deploy tests to their own workflow job

### PR DESCRIPTION
**Pull request by PaliC**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72637

Differential Revision: [D34563885](https://our.internmc.facebook.com/intern/diff/D34563885)

### CODE REVIEW

**Review by suo — CHANGES_REQUESTED** [review]
So, we need:
1. A specialized build of PyTorch, built with `USE_DEPLOY=1`.
2. A special test workflow that pretty much only calls the Python deploy tests `test/test_deploy.pt` and the C++ ones `build/bin/test_deploy`.

For the first one:
- You need to define a new workflow in `.github/scripts/generate_ci_workflow.py`. A very similar one would be `parallelnative`. You'll need to give a build environment name.
- In `.jenkins/pytorch/build.sh` you need to correctly export `USE_DEPLOY=1` if the name of the build environment matches your workflow's build environment name. Example: https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/build.sh#L60-L62

Then for the second one, I think the easiest thing to do is match on build environment name in `.jenkins/pytorch/test.sh` and call the right tests, like: https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/test.sh#L534-L536

File: .circleci/docker/build.sh [hunk 1]
```diff
@@ -336,6 +336,7 @@ docker build \
        --build-arg "KATEX=${KATEX:-}" \
        --build-arg "ROCM_VERSION=${ROCM_VERSION:-}" \
        --build-arg "PYTORCH_ROCM_ARCH=${PYTORCH_ROCM_ARCH:-gfx900;gfx906}" \
+       --build-arg "USE_DEPLOY=${USE_DEPLOY}" \
```
Comment (suo) [inline .circleci/docker/build.sh:339]: We should delete this line, it turns on deploy builds for everything. We just want it on for the deploy workflow (which we do below in build.sh anyway)

File: .github/scripts/generate_ci_workflows.py [hunk 2]
```diff
@@ -439,6 +441,15 @@ def generate_workflow_file(self, workflow_template: jinja2.Template) -> None:
             labels={LABEL_CIFLOW_LINUX, LABEL_CIFLOW_CPU},
         ),
     ),
+    CIWorkflow(
+        arch="linux",
+        build_environment="deploy-linux-xenial-cuda11.3-cudnn8-py3-gcc7",
+            docker_image_base=f"{DOCKER_REGISTRY}/pytorch/pytorch-linux-xenial-cuda11.3-cudnn8-py3-gcc7",
```
Comment (suo) [inline .github/scripts/generate_ci_workflows.py:447]: please address lint

File: .github/scripts/generate_ci_workflows.py [hunk 3]
```diff
@@ -166,6 +166,7 @@ class CIWorkflow:
     num_test_shards_on_pull_request: int = -1
     distributed_test: bool = True
     fx2trt_test: bool = False
+    deploy_test: bool = False
```
Comment (suo) [inline .github/scripts/generate_ci_workflows.py:169]: I think we don't need this; it should be controlled by the workflow name (e.g. in test.sh).

**Review by suo — CHANGES_REQUESTED** [review]
You probably need to rerun `generate_ci_workflows.py`? I don't see a generated deploy workflow. Similarly, your change should not be affecting any existing workflows, but I see several changed workflows in the PR.

**Comment by suo** [comment]
What we should be seeing is a `deploy-linux-xenial-cuda11.3-cudnn8-py3-gcc7-build` and `deploy-linux-xenial-cuda11.3-cudnn8-py3-gcc7-test` workflow, which is not in the list of signals

File: .github/scripts/generate_ci_workflows.py (hunk 3 above)
Comment (suo) [reply .github/scripts/generate_ci_workflows.py:168]: same comment

**Review by suo — CHANGES_REQUESTED** [review]

File: .github/scripts/generate_ci_workflows.py [hunk 4]
```diff
@@ -185,6 +185,7 @@ class CIWorkflow:
     enable_xla_test: bool = False
     enable_noarch_test: bool = False
     enable_force_on_cpu_test: bool = False
+    enable_deploy_test: bool = False
```
Comment (suo) [inline .github/scripts/generate_ci_workflows.py:188]: you no longer need this, it's not used anywhere.

File: .github/scripts/generate_pytorch_test_matrix.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python3
```
Comment (suo) [inline .github/scripts/generate_pytorch_test_matrix.py:1]: This file was removed, but you re-added it. Please remove

**Review by suo — CHANGES_REQUESTED** [review]

**Review by suo — APPROVED** [review]

**Comment by PaliC** [comment]
@PaliC has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34563885).

**Pull request closed without merging**